File: leetcode-solutions/ruby/121. Best Time to Buy and Sell Stock.rb
#  https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  min_price = Float::MAX
  max_profit = 0
  
  prices.each do |v|
      if v < min_price
          min_price = v
      elsif v - min_price > max_profit
          max_profit = v - min_price
      end
  end
  
  max_profit
end

File: leetcode-solutions/python/121. Best Time to Buy and Sell Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_buy_price = prices[0]
        max_profit = 0
        for p in prices:
            min_buy_price = min(p, min_buy_price)
            max_profit = max(max_profit, p - min_buy_price)
        return max_profit
